### ISSUE TITLE
Problem with autogenerated `SELECT` statement, too many columns

### ISSUE THREAD

**Issue by horkko** [kjolley/BIGSdb#328]
Hi Keith,

We've faced a problem with BIGSdb when we try to define scheme members.
Our users use the BIGSdb interface and the `BatchAdd` page to do it. 
But, after a certain amount of time, here is the error thrown by the interface:
```
The temp file containing the checked data does not exist.

Upload cannot proceed. Make sure that you haven't used the back button and are attempting to re-upload already submitted data.
```
And we can see in BIGSdb log something like:
```
Scheme.pm:147  - Check database attributes in the scheme_fields table for scheme#2 (Y.enterocolitica cgMLST)!
DBD::Pg::st execute failed: ERROR:  target lists can have at most 1664 entries at /opt/bigsdb/lib/BIGSdb/Scheme.pm line 145.

```

We can also see the automatic generated query from our PostgreSQL log:
```
...
FROM mv_scheme_2 WHERE (profile[29]=$1 OR profile[29]='N') AND (profile[30]=$2 OR profile[30]='N') AND (profile[31]=$3 OR profile[31]='N')
AND (profile[32]=$4 OR profile[32]='N') AND (profile[33]=$5 OR profile[33]='N') AND (profile[34]=$6 OR profile[34]='N') AND (profile[35]=$7 OR profile[35]='N') AND (profile[36]=$8 OR profile[36]='N')
...
AND (profile[1921]=$1921 OR profile[1921]='N');
```
It looks like the query is too long and reach the PostgreSQL limit for a `columns number` (https://www.postgresql.org/about/)

So I think it should be good to add some checks in auto-generated queries to avoid such issues

What do you think?

Emmanuel

**Comment by kjolley** [comment]
I don't quite understand what you are doing.  Could you clarify the exact steps please?  What exactly are you batch uploading? (scheme_members doesn't seem right).

This method is called when you are looking up scheme fields from within the isolate database, e.g. cgST from designations. I can see that the query will be creating a results table that includes every locus as a separate column, so if your scheme has >1664 loci then that will create a problem. The postgres column limit was one of the reasons I started using Pg arrays to store profiles in the scheme caches, but this method wil fail for the same reason. It should be possible to rewrite this so that it returns the profile as a Pg array (which only uses one column for all the alleles), but it would be helpful if you could list the exact steps needed to replicate the error.

**Comment by horkko** [comment]
Hi Keith,

Sorry I'm not clear. I'll ask exactly the user what were the steps and try to explain better.
But for now what I've read on the net is something like this https://www.mail-archive.com/pgsql-general@postgresql.org/msg118198.html
Definitely too many field in the `SELECT` statement.
I'll ask then the user if its scheme as more than `1664` loci. If so, what could be the solution to avoid such problem in the future?
Regards
Emmanuel

**Comment by kjolley** [comment]
The solution is to fix the code to not require a column for each locus value. I made a lot of changes to support hosting of schemes with >1664 loci (e.g. cgMLST) - but the Scheme::get_field_values_by_designations method seems to not have been changed (probably because our largest cgMLST scheme has 1605 loci so isn't triggering the problem). Originally the scheme cache tables (e.g. mv_scheme_2 in the seqdef database) had a column for each locus. Now they have a single 'profile' field which is an array type with the locus position in the array defined in a separate table.  The query can be changed so that instead of 'SELECT profile[0],profile[1],profile[2],field1... FROM mv_scheme_2' it will be 'SELECT profile,field1... FROM mv_scheme_2'. The method return value will need to be constructed from the profile array indirectly rather than straight from the database.

Before I do this, I need to set up a test case that is failing so that I can check it works properly.

**Comment by horkko** [comment]
Hi Keith,

I've checked the tables `scheme_members` in `isolates` table and we definitely have more than 1664 fields per schemes:
```
pubmlst_bigsdb_yersinia_seqdef=# select count(*) from scheme_members where scheme_id=2;
 count 
-------
  1727
pubmlst_bigsdb_yersinia_seqdef=# select count(*) from scheme_members where scheme_id=3;
 count 
-------
  1921
```
So as you mentioned above, it should be good to also prevent such error (too many column) to have array for column `locus`?
BTW, I've asked the user his steps to reproduce the error.
Emmanuel

**Comment by horkko** [comment]
Hi Keith,

sorry for my late reply, I've finally got the explanation from the user. Here it is:
The user goes to the `Curator's interface` and then click `++` to add `scheme members`.
500 loci belong to its `scheme1` and he want them to belong to schemes `scheme2` and `scheme3` having already respectively 1227 and 1421 loci.
To do this, he creates a file with 500 lines (3 columns (`id, loci name, loci position`)) and load it using the curator's interface.

Then the `max columns error` is thrown by Postgres.
It this explanation clear enough for you to have a test case to reproduce the problem?
Emmanuel

**Comment by kjolley** [comment]
Is this in the seqdef or isolate database?

**Comment by horkko** [comment]
Sorry, *seqdef* database

**Comment by kjolley** [comment]
Ok, thanks. I am surprised it's calling anything in Scheme.pm as that is only used by the isolate databases. I'll set this up and see what I can do.

**Comment by kjolley** [comment]
I can't replicate this at all in the seqdef database. I am able to add >1700 loci to scheme_members of a scheme with an existing primary key in both the seqdef and isolate databases (incidentally - it is much quicker to do this if you add the loci first, then the primary key field - this is due to a database trigger on scheme_members that fires on each row added but only if a primary key field exists).

As I suspected, I do get the same error when I set up the isolate database to point to this scheme and then perform a query.
